src/hardware/dac.rs

```diff
@@ -0,0 +1,279 @@
+            (Channel::Ch2, Limit::MaxV) => set_pwm(&mut self.max_v2, v_to_pwm(val)),
+            (Channel::Ch3, Limit::MaxV) => set_pwm(&mut self.max_v3, v_to_pwm(val)),
+            (Channel::Ch0, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos0, i_to_pwm(val)),
+            (Channel::Ch1, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos1, i_to_pwm(val)),
+            (Channel::Ch2, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos2, i_to_pwm(val)),
+            (Channel::Ch3, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos3, i_to_pwm(val)),
+            (Channel::Ch0, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg0, i_to_pwm(val)),
+            (Channel::Ch1, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg1, i_to_pwm(val)),
+            (Channel::Ch2, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg2, i_to_pwm(val)),
+            (Channel::Ch3, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg3, i_to_pwm(val)),
+        }
+    }
+}
+
+pub struct DacGpio {
+    pub sync0: PG3<Output<PushPull>>,
+    pub sync1: PG2<Output<PushPull>>,
+    pub sync2: PG1<Output<PushPull>>,
+    pub sync3: PG0<Output<PushPull>>,
+    pub shdn0: PG4<Output<PushPull>>,
+    pub shdn1: PG5<Output<PushPull>>,
+    pub shdn2: PG6<Output<PushPull>>,
+    pub shdn3: PG7<Output<PushPull>>,
+}
+
+/// DAC: https://www.analog.com/media/en/technical-documentation/data-sheets/AD5680.pdf
+/// Peltier Driver: https://datasheets.maximintegrated.com/en/ds/MAX1968-MAX1969.pdf
+pub struct Dac {
+    spi: Spi<SPI3, Enabled, u8>,
+    pub val: [u32; 4],
+    gpio: DacGpio,
+}
+
+impl Dac {
+    pub fn new(
+        clocks: &CoreClocks,
+        prec: rec::Spi3,
+        spi3: SPI3,
+        sck: PC10<Alternate<AF6>>,
+        mosi: PC12<Alternate<AF6>>,
+        gpio: DacGpio,
+    ) -> Self {
+        let spi = spi3.spi(
+            (sck, NoMiso, mosi),
+            MODE_1,
+            SPI_CLOCK,
+            prec,
+            clocks, // default pll1_q clock source
+        );
+
+        let mut dac = Dac {
+            spi,
+            val: [0, 0, 0, 0],
+            gpio,
+        };
+
+        dac.gpio.sync0.set_high().unwrap();
+        dac.gpio.sync1.set_high().unwrap();
+        dac.gpio.sync2.set_high().unwrap();
+        dac.gpio.sync3.set_high().unwrap();
+
+        // default to zero amps
+        dac.set(0.0, Channel::Ch0);
+        dac.set(0.0, Channel::Ch1);
+        dac.set(0.0, Channel::Ch2);
+        dac.set(0.0, Channel::Ch3);
+        dac
+    }
+
+    /// Set the DAC output to current on a channel.
+    pub fn set(&mut self, curr: f32, ch: Channel) {
+        let value = i_to_dac(curr);
+        let value = value.min(MAX_VALUE);
+        // 24 bit transfer. First 4 bit and last 2 bit are low.
+        let mut buf = [(value >> 14) as u8, (value >> 6) as u8, (value << 2) as u8];
+
+        match ch {
+            Channel::Ch0 => {
+                self.gpio.sync0.set_low().unwrap();
+                self.spi.transfer(&mut buf).unwrap();
```

Comment:
Did you check with the SPI traits API that it guarantees that the transfer is actually over by the time `transfer()` returns? Otherwise you'd be raising `sync` early.

Reply:
Proof by example is a fragile thing. May well depend on whether you are doing a release build or debug build.
Given that a [recent HAL trait](https://github.com/rust-embedded/embedded-hal/blob/47df22ca93859935df6bfc52030103a4bcece62e/src/spi/blocking.rs#L328-L329) explicitly allows returning before completion, I would check what the actual implementation is like on the HAL version that you are using.

Reply:
Ok, maybe @ryan-summers could confirm this for me: 
The stm32hal uses the [send](https://github.com/stm32-rs/stm32h7xx-hal/blob/master/src/spi.rs#L1078-L1104) function for wransfer/write which uses the [block](https://github.com/rust-embedded/nb/blob/master/src/lib.rs#L238-L267) macro and therefore the function only returns once it's done, correct?

Reply:
`send()` returns when the byte is in the FIFO. That doesn't mean anything w.r.t. transmission. Do have a look at what I said earlier.

Reply:
Oh sorry I didn't refresh the page or look in my messages so I did not see your last comment..

> The way the HAL is written, it appears that it will indeed wait for the receive to complete before returning.

Ok, I see, so the blocking read function will only return once some data has been read. Thanks.